README.md
# consul-haproxy

This project provides `consul-haproxy`, a daemon for dynamically
configuring HAProxy using data from Consul.

The daemon watches any number of backends for updates, and when
appropriate renders an configuration template for HAProxy and then
invokes a reload command, which can gracefully reload HAProxy. This
allows for a zero-downtime reload of HAProxy which is populated by
Consul.

## Download & Compilation

Download a release from the [releases page](https://github.com/hashicorp/consul-haproxy/releases), or compile from source:

```
$ make
$ ./bin/consul-haproxy -h
```

## Usage

The `consul-haproxy` command takes a number of CLI flags:

* `-addr` - Provides the HTTP address of a Consul agent. By default this
  assumes a local agent at "127.0.0.1:8500".

* `-backend` - Backend specification. Can be provided multiple times.
  The specification of a backend is documented below.

* `-dry` - Dry run. Emit config file to stdout.

* `-f` - Path to config file, overwrites CLI flags. The format of the
  file is documented below.

* `-in`- Path to a template file. This is the template that is rendered
  to generate the configuration file at `-out`. It uses the Golang templating
  system. Docs for that are [here](http://golang.org/pkg/text/template/).
  Can be provided multiple times. If specified multiple times, specify the
  same number of paths with `-out`.

* `-out` - Path to output configuration file. This path must be writable
  by `consul-haproxy` or the file cannot be updated. This can be specified
  multiple times.

* `-reload` - Command to invoke to reload configuration. This command can
  be any executable, and should be used to reload HAProxy. This is invoked
  only after the configuration file is updated.

* `-quiet` - Quiet specifies a duration of time to wait for no updates
  before writing out the new configuration. This allows for waiting until
  a service stabilizes to prevent many different reloads.

* `-max-wait` - Max wait is used to limit how waiting is done for a quiet
  period before forcing a reload. This defaults to 4x the `-quiet` value.
  As an example, if `-quiet=30s` but the backends are constantly flapping,
  a refresh will be forced after 2 minutes.

In addition to using CLI flags, `consul-haproxy` can be configured using a
file given the `-f` flag. A configuration file overrides any values given by
the CLI unless otherwise specified. The configuration file should be a JSON
object with the following keys:

* `address` - Same as `-addr` CLI flag.
* `backends` - A list of backend specifications. This is merged with any
  backends provided via the CLI.
* `dry_run` - Same as `-dry` CLI flag.
* `paths` - Same as `-out` CLI flag. . This value should be a list of paths and
  is merged with any paths provided via the CLI.
* `reload_command` - Same as `-reload` CLI flag.
* `templates` - Same as `-in` CLI flag. This value should be a list of templates
  and is merged with any paths provided via the CLI.
* `quiet` - Same as `-quiet` CLI flag.
* `max_wait` - Same as `-max-wait` CLI flag.

## Backend Specification

One of the key configuration values to `consul-haproxy` is the backends that
it should monitor. These are the service entries that are watched for changes
and used to populate the configuration file. The syntax for a backend is:

    backend_name=tag.service@datacenter:port

The specification provides a variable name for the template `backend_name`,
which is defined as the entries that match the given tag and service for a specific
datacenter. A port can also be provided, which overrides the port specified by
the service. The tag, datacenter, and port are all optional and can be omitted.

Below are a few examples:

* `app=release.webapp` - This defines a template variable `app` which watches for
  the `webapp` service, filtering on the `release` tag.

* `db=mysql@east-aws:5500` - This defines a template variable `db` which watches for
  the `mysql` service in the `east-aws` datacenter, using port 5500.

A useful features is the ability to specify multiple backends with the same variable
name. This causes the nodes to be merged. This can be used to merge nodes with various
tags, or different datacenters together. As an example, we can define:

    app=webapp@dc1
    app=webapp@dc2
    app=webapp@dc3

This backend specification sets `app` variable to be the union of the servers
in the `dc1`, `dc2`, and `dc3` datacenters.

## Template Language

The template language is the Golang text/template package, which is
[fully documented here](http://golang.org/pkg/text/template/). However, the
basic usage is quite simple.

As an example, suppose we define a simple backends with:

    app=webapp@east-aws:8000
    cache=redis


We might provide a very basic template like:

    global
        daemon
        maxconn 256

    defaults
        mode tcp
        timeout connect 5000ms
        timeout client 60000ms
        timeout server 60000ms

    listen http-in
        bind *:80{{range .app}}
        {{.}} maxconn 32{{end}}

    listen cache-in
        bind *:4444{{range .cache}}
        {{.}} maxconn 64{{end}}

This will populate the `http-in` block with the servers
in the `app` backend, and the `cache-in` with the servers
in the `cache` backend. This template will be re-rendered when
any of those servers changing, allowing for dynamic updates.

## Example

We run the example below against our
[NYC demo server](http://nyc1.demo.consul.io). This lets you set
quickly test consul-haproxy.

First lets create a simple template:

    global
        daemon
        maxconn 256

    defaults
        mode tcp
        timeout connect 5000ms
        timeout client 60000ms
        timeout server 60000ms

    listen http-in
        bind *:8000{{range .c}}
        {{.}}{{end}}

Now, we can run the following to get our output configuration:

    consul-haproxy -addr=demo.consul.io -in in.conf -backend "c=consul@nyc1:80" -backend "c=consul@sfo1:80" -dry

When this runs, we should see something like the following:

    global
        daemon
        maxconn 256

    defaults
        mode tcp
        timeout connect 5000ms
        timeout client 60000ms
        timeout server 60000ms

    listen http-in
        bind *:8000
        server 0_nyc1-consul-1_consul 192.241.159.115:80
        server 0_nyc1-consul-2_consul 192.241.158.205:80
        server 0_nyc1-consul-3_consul 198.199.77.133:80
        server 1_sfo1-consul-2_consul 162.243.155.82:80
        server 1_sfo1-consul-1_consul 107.170.195.169:80
        server 1_sfo1-consul-3_consul 107.170.195.158:80

## Varnish Example


derived from the varnish director example from the varnish page: https://www.varnish-cache.org/docs/trunk/users-guide/vcl-backends.html#directors

	import directors;    # load the directors
	{{range .c}}
	backend {{.Node}}_{{.ID}} { 
	    .host = "{{.IP}}";
	    .port = "{{.Port}}";
	}{{end}}

	sub vcl_init {
	    new bar = directors.round_robin();
	{{range .c}}
	    bar.add_backend({{.Node}}_{{.ID}});{{end}}
	}

	sub vcl_recv {
	    # send all traffic to the bar director:
	    set req.backend_hint = bar.backend();
	}

Now, we run the following command:

    consul-haproxy -addr=demo.consul.io -in in.conf -backend "c=consul@nyc1:80" -backend "c=consul@sfo1:80" -dry

The following should return:

	backend 0_nyc1-server-1_consul {
	    .host = "192.241.159.115";
	    .port = "80";
	}
	backend 0_nyc1-server-3_consul {
	    .host = "198.199.77.133";
	    .port = "80";
	}
	backend 0_nyc1-server-2_consul {
	    .host = "162.243.162.228";
	    .port = "80";
	}
	backend 1_sfo1-server-3_consul {
	    .host = "107.170.196.151";
	    .port = "80";
	}
	backend 1_sfo1-server-2_consul {
	    .host = "107.170.195.154";
	    .port = "80";
	}
	backend 1_sfo1-server-1_consul {
	    .host = "162.243.153.242";
	    .port = "80";
	}

	sub vcl_init {
	    new bar = directors.round_robin();

	    bar.add_backend(0_nyc1-server-1_consul);
	    bar.add_backend(0_nyc1-server-3_consul);
	    bar.add_backend(0_nyc1-server-2_consul);
	    bar.add_backend(1_sfo1-server-3_consul);
	    bar.add_backend(1_sfo1-server-2_consul);
	    bar.add_backend(1_sfo1-server-1_consul);
	}

	sub vcl_recv {
	    # send all traffic to the bar director:
	    set req.backend_hint = bar.backend();
	}





CHANGELOG.md
## 0.2.0 (October 09, 2014)

* Add the ability to use multiple templates & paths

## 0.1.1 (September 29, 2014)

* Support for `-quiet` and `-max-wait`
* Avoid updates on changes to notes or health check output
* Allow access to other variables in the template (thanks @charliek)

## 0.1.0 (August 11, 2014)

* Initial release


main_test.go
package main

import (
	"reflect"
	"testing"
)

func TestWatchRE(t *testing.T) {
	type match struct {
		inp   string
		parts []string
	}
	inps := []match{
		{"bar", nil},
		{"bar=", nil},
		{"=zip", nil},
		{"app=bar", []string{"app", "", "bar", "", ""}},
		{"app=tag.bar", []string{"app", "tag.", "bar", "", ""}},
		{"app=bar@dc1", []string{"app", "", "bar", "@dc1", ""}},
		{"app=bar:80", []string{"app", "", "bar", "", ":80"}},
		{"app=bar@dc1:80", []string{"app", "", "bar", "@dc1", ":80"}},
		{"app=tag.bar@dc1:80", []string{"app", "tag.", "bar", "@dc1", ":80"}},
	}

	for _, inp := range inps {
		parts := WatchRE.FindStringSubmatch(inp.inp)
		if len(parts) == 0 && len(inp.parts) != 0 {
			t.Fatalf("unexpected fail: %s", inp.inp)
		}
		if len(parts) != 0 && len(inp.parts) == 0 {
			t.Fatalf("unexpected parse: %s", inp.inp)
		}
		if len(parts) == 0 && len(inp.parts) == 0 {
			continue
		}
		if !reflect.DeepEqual(parts[1:], inp.parts) {
			t.Fatalf("bad: %v %v", parts[1:], inp.parts)
		}
	}
}

func TestReadConfig(t *testing.T) {
	conf := &Config{}
	err := readConfig("test-fixtures/config.json", conf)
	if err != nil {
		t.Fatalf("err: %v", err)
	}
	if !conf.DryRun {
		t.Fatalf("bad: %v", conf)
	}
	if conf.Address != "127.0.0.2:8500" {
		t.Fatalf("bad: %v", conf)
	}
	templates := []string{
		"test-fixtures/simple.conf",
		"test-fixtures/second.conf",
	}
	if !reflect.DeepEqual(conf.Templates, templates) {
		t.Fatalf("bad: %v", conf)
	}
	paths := []string{
		"output.conf",
		"output2.conf",
	}
	if !reflect.DeepEqual(conf.Paths, paths) {

		t.Fatalf("bad: %v", conf)
	}
	if conf.ReloadCommand != "echo 'foo' > reload_out" {
		t.Fatalf("bad: %v", conf)
	}
	backends := []string{
		"app=foo",
		"app=tag.foo",
		"app=tag.foo@dc2:8000",
	}
	if !reflect.DeepEqual(conf.Backends, backends) {
		t.Fatalf("bad: %v", conf)
	}
}

func TestValidateConfig_Good(t *testing.T) {
	conf := &Config{}
	err := readConfig("test-fixtures/config.json", conf)
	if err != nil {
		t.Fatalf("err: %v", err)
	}
	errs := validateConfig(conf)
	if len(errs) > 0 {
		t.Fatalf("err: %v", errs)
	}
	if len(conf.watches) != 3 {
		t.Fatalf("bad: %v", conf.watches)
	}
	wp1 := &WatchPath{
		Spec:    "app=foo",
		Backend: "app",
		Service: "foo",
	}
	if !reflect.DeepEqual(wp1, conf.watches[0]) {
		t.Fatalf("bad: %v", conf.watches[0])
	}
	wp2 := &WatchPath{
		Spec:    "app=tag.foo",
		Backend: "app",
		Tag:     "tag",
		Service: "foo",
	}
	if !reflect.DeepEqual(wp2, conf.watches[1]) {
		t.Fatalf("bad: %v", conf.watches[1])
	}
	wp3 := &WatchPath{
		Spec:       "app=tag.foo@dc2:8000",
		Backend:    "app",
		Tag:        "tag",
		Service:    "foo",
		Datacenter: "dc2",
		Port:       8000,
	}
	if !reflect.DeepEqual(wp3, conf.watches[2]) {
		t.Fatalf("bad: %v", conf.watches[2])
	}
}

func TestValidateConfig_Missing(t *testing.T) {
	conf := &Config{}
	errs := validateConfig(conf)
	if len(errs) != 4 {
		t.Fatalf("bad: %v", errs)
	}
}


watch_test.go
package main

import (
	"bytes"
	"github.com/armon/consul-api"
	"io/ioutil"
	"os"
	"testing"
	"time"
)

func TestMaybeRefresh(t *testing.T) {
	defer os.Remove("config_out")
	defer os.Remove("config_out2")
	defer os.Remove("reload_out")

	en1 := &consulapi.ServiceEntry{
		Node:    &consulapi.Node{Node: "node1", Address: "127.0.0.1"},
		Service: &consulapi.AgentService{ID: "app", Port: 8000},
	}
	en2 := &consulapi.ServiceEntry{
		Node:    &consulapi.Node{Node: "node3", Address: "127.0.0.3"},
		Service: &consulapi.AgentService{ID: "app", Port: 8000},
	}
	wp1 := &WatchPath{Backend: "app"}
	wp2 := &WatchPath{Backend: "app"}
	d := &backendData{
		Servers: map[*WatchPath][]*consulapi.ServiceEntry{
			wp1: []*consulapi.ServiceEntry{en1},
			wp2: []*consulapi.ServiceEntry{en2},
		},
		Backends: map[string][]*WatchPath{
			"app": []*WatchPath{wp1, wp2},
		},
	}
	templates := []string {
		"test-fixtures/simple.conf",
		"test-fixtures/varnish.vcl",
	}
	paths := []string {
		"config_out",
		"config_out2",
	}
	conf := &Config{
		watches:       []*WatchPath{wp1, wp2},
		Templates:     templates,
		Paths:         paths,
		ReloadCommand: "echo 'foo' > reload_out",
	}

	// Attempt a refresh
	if maybeRefresh(conf, d) {
		t.Fatalf("unexpected exit")
	}

	// Check config file
	out, err := ioutil.ReadFile("config_out")
	if err != nil {
		t.Fatalf("err: %v", err)
	}
	expect, err := ioutil.ReadFile("test-fixtures/simple.conf.out")
	if err != nil {
		t.Fatalf("err: %v", err)
	}
	if !bytes.Equal(out, expect) {
		t.Fatalf("bad: %s", out)
	}

	// Check second config file
	out2, err := ioutil.ReadFile("config_out2")
	if err != nil {
		t.Fatalf("err: %v", err)
	}
	expect2, err := ioutil.ReadFile("test-fixtures/varnish.vcl.out")
	if err != nil {
		t.Fatalf("err: %v", err)
	}
	if !bytes.Equal(out2, expect2) {
		t.Fatalf("bad: %s", out)
	}

	// Check reload fie
	bytes, err := ioutil.ReadFile("reload_out")
	if err != nil {
		t.Fatalf("err: %v", err)
	}
	if string(bytes) != "foo\n" {
		t.Fatalf("bad: %v", bytes)
	}
}

func TestAllWatchesReturned(t *testing.T) {
	wp1 := &WatchPath{Backend: "app"}
	wp2 := &WatchPath{Backend: "app"}
	wp3 := &WatchPath{Backend: "db"}
	d := &backendData{
		Servers: map[*WatchPath][]*consulapi.ServiceEntry{
			wp1: nil,
			wp2: nil,
		},
	}
	conf := &Config{
		watches: []*WatchPath{wp1, wp2, wp3},
	}

	if allWatchesReturned(conf, d) {
		t.Fatalf("unexpected done")
	}

	d.Servers[wp3] = nil
	if !allWatchesReturned(conf, d) {
		t.Fatalf("expected done")
	}
}

func TestAggregateServers(t *testing.T) {
	en1 := &consulapi.ServiceEntry{
		Node:    &consulapi.Node{Node: "node1", Address: "127.0.0.1"},
		Service: &consulapi.AgentService{ID: "app", Port: 8000},
	}
	en2 := &consulapi.ServiceEntry{
		Node:    &consulapi.Node{Node: "node3", Address: "127.0.0.3"},
		Service: &consulapi.AgentService{ID: "app", Port: 8000},
	}
	en3 := &consulapi.ServiceEntry{
		Node:    &consulapi.Node{Node: "node2", Address: "127.0.0.2"},
		Service: &consulapi.AgentService{ID: "db", Port: 5000},
	}
	wp1 := &WatchPath{Backend: "app"}
	wp2 := &WatchPath{Backend: "app"}
	wp3 := &WatchPath{Backend: "db"}
	d := &backendData{
		Servers: map[*WatchPath][]*consulapi.ServiceEntry{
			wp1: []*consulapi.ServiceEntry{en1},
			wp2: []*consulapi.ServiceEntry{en2},
			wp3: []*consulapi.ServiceEntry{en3},
		},
		Backends: map[string][]*WatchPath{
			"app": []*WatchPath{wp1, wp2},
			"db":  []*WatchPath{wp3},
		},
	}
	agg := aggregateServers(d)
	if len(agg) != 2 {
		t.Fatalf("Bad: %v", agg)
	}
	app := agg["app"]
	if len(app) != 2 {
		t.Fatalf("Bad: %v", app)
	}
	if app[0] != en1 && app[1] != en2 {
		t.Fatalf("Bad: %v", app)
	}
	db := agg["db"]
	if len(db) != 1 {
		t.Fatalf("Bad: %v", db)
	}
	if db[0] != en3 {
		t.Fatalf("Bad: %v", db)
	}
}

func TestBuildTemplate(t *testing.T) {
	templates := []string {
		"test-fixtures/simple.conf",
		"test-fixtures/varnish.vcl",
	}
	expectations := []string {
		"test-fixtures/simple.conf.out",
		"test-fixtures/varnish.vcl.out",
	}
	servers := map[string][]*consulapi.ServiceEntry{
		"app": []*consulapi.ServiceEntry{
			&consulapi.ServiceEntry{
				Node:    &consulapi.Node{Node: "node1", Address: "127.0.0.1"},
				Service: &consulapi.AgentService{ID: "app", Port: 8000},
			},
			&consulapi.ServiceEntry{
				Node:    &consulapi.Node{Node: "node3", Address: "127.0.0.3"},
				Service: &consulapi.AgentService{ID: "app", Port: 8000},
			},
		},
	}

	// Iterate through the list of templates to render
	for idx, templatePath := range templates {
		out, err := buildTemplate(templatePath, servers)
		if err != nil {
			t.Fatalf("err: %v", err)
		}

		expect, err := ioutil.ReadFile(expectations[idx])
		if err != nil {
			t.Fatalf("err: %v", err)
		}

		if !bytes.Equal(out, expect) {
			t.Fatalf("bad: %s", out)
		}

	}
}

func TestReload(t *testing.T) {
	os.Remove("test_out")
	conf := &Config{
		ReloadCommand: "echo 'foo' > test_out",
	}
	if err := reload(conf); err != nil {
		t.Fatalf("err: %v", err)
	}
	bytes, err := ioutil.ReadFile("test_out")
	if err != nil {
		t.Fatalf("err: %v", err)
	}
	if string(bytes) != "foo\n" {
		t.Fatalf("bad: %v", bytes)
	}
	os.Remove("test_out")
}

func TestShouldStop(t *testing.T) {
	ch := make(chan struct{})
	if shouldStop(ch) {
		t.Fatalf("bad")
	}
	close(ch)
	if !shouldStop(ch) {
		t.Fatalf("bad")
	}
}

func TestAsyncNotify(t *testing.T) {
	ch := make(chan struct{}, 1)
	asyncNotify(ch)
	asyncNotify(ch)
	asyncNotify(ch)

	select {
	case <-ch:
	default:
		t.Fatalf("should work")
	}
	select {
	case <-ch:
		t.Fatalf("should not work")
	default:
	}

}

func TestMin(t *testing.T) {
	if min(1, 2) != 1 {
		t.Fatalf("Bad")
	}
	if min(2, 1) != 1 {
		t.Fatalf("Bad")
	}
}

func TestBackoff(t *testing.T) {
	type val struct {
		fail   int
		expect time.Duration
	}
	inps := []val{
		{1, 3 * time.Second},
		{2, 6 * time.Second},
		{3, 12 * time.Second},
		{4, 24 * time.Second},
	}
	for _, inp := range inps {
		if out := backoff(3*time.Second, inp.fail); out != inp.expect {
			t.Fatalf("bad: %v %v", inp, out)
		}
	}
}

func TestFormatOutput(t *testing.T) {
	inp := map[string][]*consulapi.ServiceEntry{
		"foo": []*consulapi.ServiceEntry{
			&consulapi.ServiceEntry{
				Node:    &consulapi.Node{Node: "node1", Address: "127.0.0.1"},
				Service: &consulapi.AgentService{ID: "redis", Port: 8000},
			},
			&consulapi.ServiceEntry{
				Node:    &consulapi.Node{Node: "node3", Address: "127.0.0.3"},
				Service: &consulapi.AgentService{ID: "redis", Port: 1234},
			},
		},
		"bar": []*consulapi.ServiceEntry{
			&consulapi.ServiceEntry{
				Node:    &consulapi.Node{Node: "node2", Address: "127.0.0.2"},
				Service: &consulapi.AgentService{ID: "memcache", Port: 80},
			},
			&consulapi.ServiceEntry{
				Node:    &consulapi.Node{Node: "node4", Address: "127.0.0.4"},
				Service: &consulapi.AgentService{ID: "memcache", Port: 10000},
			},
		},
	}

	output := formatOutput(inp)
	if len(output) != 2 {
		t.Fatalf("bad: %v", output)
	}
	foo := output["foo"]
	if len(foo) != 2 {
		t.Fatalf("bad: %v", foo)
	}
	if foo[0].String() != "server node1_redis 127.0.0.1:8000" {
		t.Fatalf("Bad: %v", foo)
	}
	if foo[1].String() != "server node3_redis 127.0.0.3:1234" {
		t.Fatalf("Bad: %v", foo)
	}

	bar := output["bar"]
	if len(bar) != 2 {
		t.Fatalf("bad: %v", bar)
	}
	if bar[0].String() != "server node2_memcache 127.0.0.2:80" {
		t.Fatalf("Bad: %v", bar)
	}
	if bar[1].String() != "server node4_memcache 127.0.0.4:10000" {
		t.Fatalf("Bad: %v", bar)
	}
}


main.go
package main

import (
	"bytes"
	"encoding/json"
	"errors"
	"flag"
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"os/signal"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"syscall"
	"time"

	"github.com/mitchellh/mapstructure"
)

// WatchRE is used to parse a backend configuration. The config should
// look like "backend=tag.service@datacenter:port". However, the tag, port and
// datacenter are optional, so it can also be provided as "backend=service"
var WatchRE = regexp.MustCompile("^([^=]+)=([^.]+\\.)?([^.:@]+)(@[^.:]+)?(:[0-9]+)?$")

// WatchPath represents a path we need to watch
type WatchPath struct {
	Spec       string
	Backend    string
	Service    string
	Tag        string
	Datacenter string
	Port       int
}

// Config is used to configure the HAProxy connector
type Config struct {
	// DryRun is used to avoid actually modifying the file
	// or reloading HAProxy.
	DryRun bool `mapstructure:"dry_run"`

	// Address is the Consul HTTP API address
	Address string `mapstructure:"address"`

	// Path to the HAProxy template file
	Templates []string `mapstructure:"templates"`

	// Path to the HAProxy configuration file to write
	Paths []string `mapstructure:"paths"`

	// Command used to reload HAProxy
	ReloadCommand string `mapstructure:"reload_command"`

	// Backends are used to specify what we watch. Given as:
	// "name=(tag.)service"
	Backends []string `mapstructure:"backends"`

	// Quiet is how long we wait for a "quiet" period before
	// trigger the re-build and re-load. This allows us to
	// wait for the system to reach a quiescent state instead
	// of trigger constant updates.
	Quiet time.Duration `mapstructure:"quiet"`

	// MaxWait works with Quiet to limit the maximum amount of
	// time we wait before forcing a reload to take place.
	// In the case of an unstable system (constant flapping),
	// this allows progress to be made. Defaults to 4x the
	// Quiet value if not provided.
	MaxWait time.Duration `mapstructure:"max_wait"`

	// watches are the watches we need to track
	watches []*WatchPath
}

func main() {
	os.Exit(realMain())
}

// getConfig is used to read our configuration
func getConfig() (*Config, error) {
	var configFile string
	var backends []string
	var templates  []string
	var paths []string

	conf := &Config{}
	cmdFlags := flag.NewFlagSet("consul-haproxy", flag.ContinueOnError)
	cmdFlags.Usage = usage
	cmdFlags.StringVar(&conf.Address, "addr", "127.0.0.1:8500", "consul HTTP API address with port")
	cmdFlags.Var((*AppendSliceValue)(&templates), "in", "template path")
	cmdFlags.Var((*AppendSliceValue)(&paths), "out", "config path")
	cmdFlags.StringVar(&conf.ReloadCommand, "reload", "", "reload command")
	cmdFlags.StringVar(&configFile, "f", "", "config file")
	cmdFlags.BoolVar(&conf.DryRun, "dry", false, "dry run")
	cmdFlags.DurationVar(&conf.Quiet, "quiet", 0, "quiet period")
	cmdFlags.DurationVar(&conf.MaxWait, "max-wait", 0, "maximum wait for a quiet period")
	cmdFlags.Var((*AppendSliceValue)(&backends), "backend", "backend to populate")
	if err := cmdFlags.Parse(os.Args[1:]); err != nil {
		return nil, err
	}

	// Parse the configuration file if given
	if configFile != "" {
		if err := readConfig(configFile, conf); err != nil {
			return nil, fmt.Errorf("Failed to read config file: %v", err)
		}
	}

	// Merge the templates, paths, and backends together
	conf.Templates = append(conf.Templates, templates...)
	conf.Paths = append(conf.Paths, paths...)
	conf.Backends = append(conf.Backends, backends...)
	return conf, nil
}

// realMain is the actual entry point, but we wrap it to set
// a proper exit code on return
func realMain() int {
	if len(os.Args) == 1 {
		usage()
		return 1
	}

	// Read the configuration
	conf, err := getConfig()
	if err != nil {
		log.Printf("[ERR] %v", err)
		return 1
	}

	// Sanity check the configuration
	if errs := validateConfig(conf); len(errs) != 0 {
		for _, err := range errs {
			log.Printf("[ERR] %v", err)
		}
		return 1
	}

	// Start watching for changes
	stopCh, finishCh := watch(conf)

	// Wait for termination
	return waitForTerm(conf, stopCh, finishCh)
}

// readConfig is used to read a configuration file
func readConfig(path string, config *Config) error {
	// Read the file
	contents, err := ioutil.ReadFile(path)
	if err != nil {
		return err
	}

	// Decode the file
	var raw interface{}
	if err := json.NewDecoder(bytes.NewReader(contents)).Decode(&raw); err != nil {
		return err
	}

	// Map to our output
	if err := mapstructure.Decode(raw, config); err != nil {
		return err
	}
	return nil
}

// validateConfig is used to sanity check the configuration
func validateConfig(conf *Config) (errs []error) {
	// Check the template
	if len(conf.Templates) == 0 {
		errs = append(errs, errors.New("missing template path"))
	} else {
		for _, t := range conf.Templates {
			_, err := ioutil.ReadFile(t)
			if err != nil {
				errs = append(errs, fmt.Errorf("failed to read template '%s': %v", t, err))
			}
		}
	}

	if len(conf.Paths) == 0 && !conf.DryRun {
		errs = append(errs, errors.New("missing configuration path"))
	}

	if (len(conf.Templates) != len(conf.Paths)) && !conf.DryRun {
		errs = append(errs, errors.New("number of templates and paths do not match"))
	}

	if conf.ReloadCommand == "" && !conf.DryRun {
		errs = append(errs, errors.New("missing reload command"))
	}

	if len(conf.Backends) == 0 {
		errs = append(errs, errors.New("missing backends to populate"))
	}

	for _, b := range conf.Backends {
		parts := WatchRE.FindStringSubmatch(b)
		if parts == nil || len(parts) != 6 {
			errs = append(errs, fmt.Errorf("Backend '%s' could not be parsed", b))
			continue
		}
		var port int
		if parts[5] != "" {
			p, err := strconv.ParseInt(strings.TrimPrefix(parts[5], ":"), 10, 64)
			if err != nil {
				errs = append(errs, fmt.Errorf("Backend '%s' port could not be parsed", b))
				continue
			}
			port = int(p)
		}
		wp := &WatchPath{
			Spec:       parts[0],
			Backend:    parts[1],
			Tag:        strings.TrimSuffix(parts[2], "."),
			Service:    parts[3],
			Datacenter: strings.TrimPrefix(parts[4], "@"),
			Port:       port,
		}
		conf.watches = append(conf.watches, wp)
	}

	// Ensure a non-negative time interval
	if conf.Quiet < 0 || conf.MaxWait < 0 {
		errs = append(errs, errors.New("Cannot specify a negative time interval"))
	}

	// Handle setting a default MaxWait if a quiet period
	// is configured
	if conf.Quiet != 0 && conf.MaxWait == 0 {
		conf.MaxWait = 4 * conf.Quiet
	}

	return
}

// waitForTerm waits until we receive a signal to exit
func waitForTerm(conf *Config, stopCh, finishCh chan struct{}) int {
	signalCh := make(chan os.Signal, 1)
	signal.Notify(signalCh, os.Interrupt, syscall.SIGHUP)
	for {
		select {
		case sig := <-signalCh:
			switch sig {
			case syscall.SIGHUP:
				// Read the configuration
				log.Printf("[INFO] SIGHUP received, reloading configuration...")
				newConf, err := getConfig()
				if err != nil {
					log.Printf("[ERR] Failed to read new config: %v", err)
					continue
				}

				// Sanity check the configuration
				if errs := validateConfig(newConf); len(errs) != 0 {
					for _, err := range errs {
						log.Printf("[ERR] %v", err)
					}
					continue
				}

				// Switch to the new configuration
				conf = newConf

				// Stop the existing watcher
				close(stopCh)

				// Start a new watcher
				stopCh, finishCh = watch(conf)
				log.Printf("[INFO] Configuration reload complete")

			default:
				log.Printf("[WARN] Received %v signal, shutting down", sig)
				return 0
			}
		case <-finishCh:
			if conf.DryRun {
				return 0
			}
			log.Printf("[WARN] Aborting watching for changes, shutting down")
			return 1
		}
	}
	return 0
}

func usage() {
	cmd := filepath.Base(os.Args[0])
	fmt.Fprintf(os.Stderr, strings.TrimSpace(helpText)+"\n\n", cmd)
}

const helpText = `
Usage: %s [options]

  Watches a service group in Consul and dynamically configures
  an HAProxy backend. The process runs continuously, monitoring
  all the backends for changes. When there is a change, the template
  file is rendered to a destination path, and a reload command is
  invoked. This allows HAProxy configuration to be updated in real
  time using Consul.

  Backends are specified using the following syntax:

    app=release.webapp@east-aws:8000

  In this syntax, we are defining a template variable 'app',
  which is populated from the 'webapp' service, 'release' tag, in the
  'east-aws' datacenter, using port 8000. If the port is given it
  overrides any specified by the service. The tag, datacenter
  and port are optional. So we could also specify this as:

    app=webapp

  This exports the 'app' variable as just the nodes in the 'webapp'
  service in the local datacenter. Multiple backends can be specified,
  and even multiple watches for a given backend.

  For example:

    app=webapp@east-aws
    app=webapp@west-aws

  This will watch both the 'east-aws' and 'west-aws' datacenters to
  populate the nodes in the 'app' backend. This can be used to merge
  multiple tags, datacenters, etc into a single backend.

Options:

  -addr=127.0.0.1:8500  Provides the HTTP address of a Consul agent.
  -backend=spec         Backend specification. Can be provided multiple times.
  -dry                  Dry run. Emit config file to stdout.
  -f=path               Path to config file, overwrites CLI flags
  -in=path              Path to a template file.  Can be provided multiple times.
  -out=path             Path to output configuration file. Can be provided multiple times.
  -reload=cmd           Command to invoke to reload configuration
  -quiet=0s             Period to wait without updates before trigger reload.
  -max-wait=0s          Maxium time to wait for quiet period. Default 4x of -quiet.
`
